pkg/logql/sharding.go

```diff
@@ -55,3 +65,327 @@ func (c ConcatLogSelectorExpr) String() string {
 
 	return fmt.Sprintf("%s ++ %s", c.LogSelectorExpr.String(), c.next.String())
 }
+
+type Shards []astmapper.ShardAnnotation
+
+func (xs Shards) Encode() (encoded []string) {
+	for _, shard := range xs {
+		encoded = append(encoded, shard.String())
+	}
+
+	return encoded
+}
+
+// ParseShards parses a list of string encoded shards
+func ParseShards(strs []string) (Shards, error) {
+	if len(strs) == 0 {
+		return nil, nil
+	}
+	shards := make([]astmapper.ShardAnnotation, 0, len(strs))
+
+	for _, str := range strs {
+		shard, err := astmapper.ParseShard(str)
+		if err != nil {
+			return nil, err
+		}
+		shards = append(shards, shard)
+	}
+	return shards, nil
+}
+
+// Downstreamer is an interface for deferring responsibility for query execution.
+// It is decoupled from but consumed by a downStreamEvaluator to dispatch ASTs.
+type Downstreamer interface {
+	Downstream(Expr, Params, Shards) (Query, error)
+}
+
+// DownstreamEvaluator is an evaluator which handles shard aware AST nodes
+type DownstreamEvaluator struct {
+	Downstreamer
+	defaultEvaluator *DefaultEvaluator
+}
+
+// Exec runs a query and collects stats from the embedded Downstreamer
+func (ev DownstreamEvaluator) Exec(ctx context.Context, expr Expr, p Params, shards Shards) (Result, error) {
+	qry, err := ev.Downstream(expr, p, shards)
+	if err != nil {
+		return Result{}, err
+	}
+
+	res, err := qry.Exec(ctx)
+	if err != nil {
+		return Result{}, err
+	}
+
+	err = stats.JoinResults(ctx, res.Statistics)
+	if err != nil {
+		level.Warn(util.Logger).Log("msg", "unable to merge downstream results", "err", err)
+	}
+
+	return res, nil
+
+}
+
+func NewDownstreamEvaluator(downstreamer Downstreamer) *DownstreamEvaluator {
+	return &DownstreamEvaluator{
+		Downstreamer: downstreamer,
+		defaultEvaluator: NewDefaultEvaluator(
+			QuerierFunc(func(_ context.Context, p SelectParams) (iter.EntryIterator, error) {
+				// TODO(owen-d): add metric here, this should never happen.
+				return nil, errors.New("Unimplemented")
+			}),
+			0,
+		),
+	}
+}
+
+// Evaluator returns a StepEvaluator for a given SampleExpr
+func (ev *DownstreamEvaluator) StepEvaluator(
+	ctx context.Context,
+	nextEv Evaluator,
+	expr SampleExpr,
+	params Params,
+) (StepEvaluator, error) {
+	switch e := expr.(type) {
+
+	case DownstreamSampleExpr:
+		// downstream to a querier
+		var shards []astmapper.ShardAnnotation
+		if e.shard != nil {
+			shards = append(shards, *e.shard)
+		}
+		res, err := ev.Exec(ctx, e.SampleExpr, params, shards)
+		if err != nil {
+			return nil, err
+		}
+		return ResultStepEvaluator(res, params)
+
+	case *ConcatSampleExpr:
+		type result struct {
+			stepper StepEvaluator
+			err     error
+		}
+		ctx, cancel := context.WithCancel(ctx)
+		cur := e
+		ch := make(chan result)
+		done := make(chan struct{})
+		count := 0
+
+		for cur != nil {
+			go func(expr SampleExpr) {
+				eval, err := ev.StepEvaluator(ctx, nextEv, expr, params)
+				select {
```

Comment:
adding a log if there's an error might help catch swallowed errors by the context cancellation. May be a warning.

Reply:
Adding this, but it's a recursive call so we'll see a ton of spans when this happens (roughly shard_factor * AST_depth). However it should be a rare occurrence and the verbosity should help.

Reply:
Can't you use util.Logger instead of a span logger ?